gtsam/hybrid/HybridGaussianFactorGraph.cpp

```diff
@@ -344,68 +344,71 @@ EliminateHybrid(const HybridGaussianFactorGraph &factors,
   // However this is also the case with iSAM2, so no pressure :)
 
   // PREPROCESS: Identify the nature of the current elimination
-  std::unordered_map<Key, DiscreteKey> mapFromKeyToDiscreteKey;
-  std::set<DiscreteKey> discreteSeparatorSet;
-  std::set<DiscreteKey> discreteFrontals;
 
+  // First, identify the separator keys, i.e. all keys that are not frontal.
   KeySet separatorKeys;
-  KeySet allContinuousKeys;
-  KeySet continuousFrontals;
-  KeySet continuousSeparator;
-
-  // This initializes separatorKeys and mapFromKeyToDiscreteKey
   for (auto &&factor : factors) {
     separatorKeys.insert(factor->begin(), factor->end());
+  }
+  // remove frontals from separator
+  for (auto &k : frontalKeys) {
+    separatorKeys.erase(k);
+  }
+
+  // Build a map from keys to DiscreteKeys
+  std::unordered_map<Key, DiscreteKey> mapFromKeyToDiscreteKey;
+  for (auto &&factor : factors) {
     if (!factor->isContinuous()) {
       for (auto &k : factor->discreteKeys()) {
         mapFromKeyToDiscreteKey[k.first] = k;
       }
     }
   }
 
-  // remove frontals from separator
-  for (auto &k : frontalKeys) {
-    separatorKeys.erase(k);
-  }
-
-  // Fill in discrete frontals and continuous frontals for the end result
+  // Fill in discrete frontals and continuous frontals.
+  std::set<DiscreteKey> discreteFrontals;
+  KeySet continuousFrontals;
   for (auto &k : frontalKeys) {
     if (mapFromKeyToDiscreteKey.find(k) != mapFromKeyToDiscreteKey.end()) {
       discreteFrontals.insert(mapFromKeyToDiscreteKey.at(k));
     } else {
       continuousFrontals.insert(k);
-      allContinuousKeys.insert(k);
     }
   }
 
-  // Fill in discrete frontals and continuous frontals for the end result
+  // Fill in discrete discrete separator keys and continuous separator keys.
+  std::set<DiscreteKey> discreteSeparatorSet;
+  KeySet continuousSeparator;
   for (auto &k : separatorKeys) {
     if (mapFromKeyToDiscreteKey.find(k) != mapFromKeyToDiscreteKey.end()) {
       discreteSeparatorSet.insert(mapFromKeyToDiscreteKey.at(k));
     } else {
       continuousSeparator.insert(k);
-      allContinuousKeys.insert(k);
     }
   }
 
-  // NOTE: We should really defer the product here because of pruning
+  // Check if we have any continuous keys:
+  const bool discrete_only =
+      continuousFrontals.empty() && continuousSeparator.empty();
 
-  // Case 1: we are only dealing with continuous
-  if (mapFromKeyToDiscreteKey.empty() && !allContinuousKeys.empty()) {
-    return continuousElimination(factors, frontalKeys);
-  }
+  // NOTE: We should really defer the product here because of pruning
 
-  // Case 2: we are only dealing with discrete
-  if (allContinuousKeys.empty()) {
+  if (discrete_only) {
+    // Case 1: we are only dealing with discrete
     return discreteElimination(factors, frontalKeys);
-  }
-
+  } else {
```

Comment:
Any particular reason for this nesting? Even an `else if` will only evaluate if `discrete_only` is false.